### PR TITLE
Additional volume with own specified temperature in cavity pressure

### LINKED ISSUE #15011: Add additional volume capability in the Cavity Pressure calculation
## Reason

Some furnace  experiments that measure cavity pressures have free volume that is outside of the specimen (e.g., tubing) that is in direct communication with the specimen. Currently, you can add an addition into the `InternalVolume` postprocessor that usually feeds into `CavityPressure`, which gives the correct volume of gas, but since this additional volume is outside of the heated area the temperature of the gas is overpredicted resulting in an overprediction of the cavity pressure.  

Currently, to obtain the correct cavity pressure, the specimen is made longer such that the calculated internal volume is correct without using the addition parameter, as well as adjusting the temperature boundary condition such that the temperature is cooler in this additional volume area. However, the problem is making a longer specimen to obtain correct pressure calculations causes incorrect solutions in the mechanics because now the specimen in some cases is twice as long as it is in the experiment.

## Design

Add the capability to specify an additional volume that has its own specified temperature which is not part of the mesh that directly communicates with the enclosed cavity for the purposes of the pressure calculation.   The `temperature` of the total volume will be a volume weighted averaged of the cavity and the additional volume such that the cavity pressure calculation becomes:
 
`  p = n * R * (w_1 * T + w_2 * T_add) / (V_tot)
`
where
   p is the pressure
   n is the amount of material in the volume (moles)
   R is the universal gas constant
   T is the temperature in the cavity
   T_add is the temperature of the additional volume
   V_tot is the volume of the cavity plus the additional volume: V_tot = V_cavity + V_add
      where V_cavity is the volume of the cavity and V_add is the additional volume 
   w_1 and w_2 are weighting values:
     w_1 = V_cavity / V_tot and w_2 = V_add / V_tot

## Impact

New parameters in an existing class.  Results of all existing tests will be the same. A new test will be added to demonstrate capability.

### PR DESCRIPTION
**Pull request by gambka**
Closes #15011.

### CODE REVIEW

**Comment by gambka** [comment]
I'm going to extend the capability to allow for multiple additional volumes with respective temperatures as per discussion with @jasondhales

**Comment by gambka** [comment]
The test failing threads on Mac test is unrelated to the changes made here.

File: modules/tensor_mechanics/doc/content/source/userobjects/CavityPressureUserObject.md [hunk 1]
```diff
@@ -21,6 +21,14 @@ original pressure, temperature, and volume) plus the amount in the cavity due to
 during the simulation.  The volume $V$ is provided by a single postprocessor value or a vector of
 postprocessor values that are summed together before being used in [eq:ideal_gas_law].
 
+There is the option to specify additional volumes with corresponding temperatures are not meshed,
+but communicate directly with the cavity. In this case the pressure becomes:
```
Comment (jasondhales) [inline modules/tensor_mechanics/doc/content/source/userobjects/CavityPressureUserObject.md:25]: There is an option to specify additional, unmeshed volumes with corresponding temperatures that communicate directly with the cavity.

File: modules/tensor_mechanics/src/userobjects/CavityPressureUserObject.C [hunk 2]
```diff
@@ -65,6 +81,24 @@ CavityPressureUserObject::CavityPressureUserObject(const InputParameters & param
...
   for (unsigned int i = 0; i < volume_names.size(); ++i)
     _volume[i] = &getPostprocessorValueByName(volume_names[i]);
+
+  if (isParamValid("additional_volumes") && !isParamValid("temperature_of_additional_volumes"))
+    mooseError("If additional volumes are specified, the temperatures of the volumes must also be "
+               "specified.");
+
+  if (!isParamValid("additional_volumes") && isParamValid("temperature_of_additional_volumes"))
+    mooseError(
+        "Additional volumes must be specified when the temperatures of the volumes are specified.");
```
Comment (jasondhales) [inline modules/tensor_mechanics/src/userobjects/CavityPressureUserObject.C:91]: You could have the check be (isPV(volume) != isPV(temp)) mooseError(...);  That would check both cases at once.  Just a comment.  Change if you like.

Comment (gambka) [reply modules/tensor_mechanics/src/userobjects/CavityPressureUserObject.C:91]: I like that change.

**Review by jasondhales — APPROVED** [review]

**Comment by gambka** [comment]
@lindsayad, On moosebuild.org the heavy jacobian test for AD that I added seems to be failing heavy testing. I did notice I missed AD for two of the FunctionDirchletBC's. Adding AD to those two FunctionDirichletBCs have not changed the behavior and I'm still getting the incorrect Jacobian. @tophmatthews, you originally worked on the AD implementation of this CavityPressure object, maybe you might have an idea too.

**Comment by lindsayad** [comment]
Did this Jacobian test ever pass for you on your local machine? So the problem is that AD is not going to help with postprocessor values that are functions of the degrees of freedom. AD is fairly expensive as is without considering dense coupling, e.g. a postprocessor could be a function of every dof in the nonlinear system. So I'm actually going to remove the Jacobian test

**Comment by gambka** [comment]
It did when I had the old version of the code that was a single real value for addition and the temperature.  But I was asked to change to a vector of postprocessors instead, and I forgot to rerun it as it was a heavy test.  You are fine to remove it.  Also the postprocessors in these tests are simply constant functions so they do not depend on the degrees of freedom, but I agree it could be.

**Comment by lindsayad** [comment]
I just briefly looked through all the CavityPressure objects and I thought I saw some calculations depending on the average temperature postprocessor...was I mistaken?

**Comment by gambka** [comment]
Yes that is correct, but that coupling is done in all the other AD tests in that directory and they seem to give perfect Jacobians in their tests.  I've just added two new vectors of postprocessors that can be inputs that do factor into the calculation if they are provided.

**Comment by lindsayad** [comment]
Yea I guess I dunno if I have an explanation for that particular test, but in general that stated requirement for that test is false because of the existence of the postprocessor, so I think it makes sense to remove

**Comment by gambka** [comment]
My comment is, that all of the Cavity Pressure calculations are based off of postprocessors on the volume and the temperature (even before my changes here), so why do the other Jacobian tests pass fine? In general the requirements for all those Jacobian tests would be false.

**Comment by lindsayad** [comment]
You're right. I don't have anything else to say then that maybe this PR has a bug in it

**Comment by lindsayad** [comment]
I guess you didn't create that mesh, @tophmatthews did. Basically if you want me to look into the Jacobian  more, I'll need a nice small test

**Comment by gambka** [comment]
I think the mesh came from the non-ad tests which were added a long time ago. I do not know where the .jou file would be.

**Comment by tophmatthews** [comment]
> I do not know where the .jou file would be.

Yeah, I should have changed that to a generated mesh, I wonder if there was a reason I didn't beyond consistency with the non-AD tests. Probably laziness.

**Pull request merged**